### PR TITLE
Make JBMC commandline 'java' compatible [TG-9888]

### LINKED ISSUE #2864: Make jbmc classpath behaviour `java`-compliant
JBMC should behave like the `java`-executable. Currently there are many deviations as https://github.com/diffblue/cbmc/pull/2733 shows.

### PR DESCRIPTION
**Pull request by peterschrammel**
As a user, I'd like to use `jbmc` in a way similar to `java`.
For example, I'd like to run `java org.cprover.MyApp` resp. `jbmc org.cprover.Myapp`.
Similarly, I'd like to run `java -jar myapp.jar` resp. `jbmc -jar myapp.jar`.
This PR brings `jbmc`, `janalyzer` and `jdiff` a step closer to that: the most common combinations of `-jar` and `-cp` are now supported.

Fixes #2864.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [X] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1d35ecdea256f82c332f2351f027ddb851ada3d3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114658916
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3898da02ea8cf6eab64b9dce5c439c7373768c4b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114722731

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c9e6b0c97bc86c06643a5dc79e8b5b9418db734d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114875345
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 916aea1a98961ed214da52e3d15e9dacf7a45774).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114875958

**Comment by peterschrammel** [comment]
TG is passing.

File: src/goto-programs/lazy_goto_model.cpp [hunk 1]
```diff
@@ -134,7 +134,35 @@ void lazy_goto_modelt::initialize(
       sources.push_back(file);
   }
 
-  if(!sources.empty())
+  if(sources.empty() && !config.java.main_class.empty())
+  {
+    // We assume it's Java.
+    const std::string filename = "";
+    language_filet &lf = add_language_file(filename);
+    lf.language = get_language_from_mode(ID_java);
```
Comment (smowton) [inline src/goto-programs/lazy_goto_model.cpp]: Perhaps this option would be better taken as a constructor arg?

File: src/goto-programs/lazy_goto_model.cpp [hunk 2]
```diff
@@ -134,7 +134,35 @@ void lazy_goto_modelt::initialize(
...
+    CHECK_RETURN(lf.language != nullptr);
+
+    languaget &language = *lf.language;
+    language.set_message_handler(message_handler);
+    language.set_language_options(options);
+
+    msg.status() << "Parsing ..." << messaget::eom;
+
+    std::ifstream infile(filename);
+    // Both arguments to parse() are discarded.
```
Comment (smowton) [inline src/goto-programs/lazy_goto_model.cpp]: This is specific to Java I suppose -- perhaps we should give `languaget` an alternative init method to be used when it doesn't represent a single file?

**Review by smowton — APPROVED** [review]

File: src/goto-programs/lazy_goto_model.cpp (hunk 2 above)
Comment (peterschrammel) [reply src/goto-programs/lazy_goto_model.cpp]: Yes, when no file is given then we assume that it is Java. There's no file extension we could look at to determine the language.

Comment (smowton) [reply src/goto-programs/lazy_goto_model.cpp]: I mean shouldn't the driver program tell the lazy-goto-model what it's expecting, and then call some other `languaget` entry-point instead of `parse` with a fake parameter?

File: regression/test.pl [hunk 3]
```diff
@@ -33,7 +33,12 @@ sub with_color {
 
 sub run($$$$$) {
   my ($name, $input, $cmd, $options, $output) = @_;
-  my $cmdline = "$cmd $options '$input' >'$output' 2>&1";
+  my $cmdline;
```
Comment (thk123) [inline regression/test.pl]: :pick: Typo in commit message for commit `Do not pass empty commandline argument in test.pl `: 
esample -> example

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 4]
```diff
@@ -1029,6 +1029,13 @@ void jbmc_parse_optionst::help()
     "                              manifest (checked in this order), the behavior is\n" // NOLINT(*)
     "                              the same as running jbmc on the corresponding\n" // NOLINT(*)
     "                              class file."
+    " jbmc -jar jarfile            JAR file to be checked\n"
```
Comment (thk123) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: :pick: As per guidelines, these help strings should be put in a macro to avoid copy pasting the same info :rabbit:

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 5]
```diff
@@ -694,6 +694,13 @@ void janalyzer_parse_optionst::help()
     "                              manifest (checked in this order), the behavior is\n" // NOLINT(*)
     "                              the same as running janalyzer on the corresponding\n" // NOLINT(*)
     "                              class file."
+    " janalyzer -jar jarfile       JAR file to be checked\n"
```
Comment (thk123) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: :rabbit:

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 6]
```diff
@@ -700,6 +700,12 @@ void janalyzer_parse_optionst::help()
     "                              of the class specified by --main-class or the main\n" // NOLINT(*)
     "                              class specified in the JAR manifest\n"
     "                              (checked in this order).\n"
+    " janalyzer --gb goto-binary   goto-binary file to be checked\n"
```
Comment (thk123) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: :rabbit2:

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 7]
```diff
@@ -1035,6 +1035,12 @@ void jbmc_parse_optionst::help()
     "                              of the class specified by --main-class or the main\n" // NOLINT(*)
     "                              class specified in the JAR manifest\n"
     "                              (checked in this order).\n"
+    " jbmc --gb goto-binary        goto-binary file to be checked\n"
```
Comment (thk123) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: :rabbit2:

File: jbmc/regression/janalyzer-taint/taint-aliasing1/test.desc [hunk 8]
```diff
@@ -1,5 +1,5 @@
 CORE
-aliasing1.class
+aliasing1
```
Comment (thk123) [inline jbmc/regression/janalyzer-taint/taint-aliasing1/test.desc:2]: :no_entry_sign: The commit message does not match with this change
` Make janalyzer use lazy goto model initialization`

File: jbmc/regression/janalyzer/too-many-args/test.desc [hunk 9]
```diff
@@ -1,6 +1,6 @@
 CORE
-A.class
-B.class
-^Only one \.class, \.jar or \.gbf file should be directly specified on the command-line
+A
+B
+Please give exactly one class name, and/or use -jar jarfile or --gb goto-binary
```
Comment (thk123) [inline jbmc/regression/janalyzer/too-many-args/test.desc:4]: I'd put this in a different commit to ease testing

File: jbmc/regression/jbmc/classpath-class-incorrect-package/test.desc [hunk 10]
```diff
@@ -1,11 +1,9 @@
-KNOWNBUG
-Test.class
+CORE
```
Comment (thk123) [inline jbmc/regression/jbmc/classpath-class-incorrect-package/test.desc:1]: :no_entry_sign: This absolutely should be in its own commit

File: jbmc/regression/jbmc/classpath-class-with-jar/test.desc [hunk 11]
```diff
@@ -1,7 +1,8 @@
 CORE
-Test.class
+Test
 --classpath `../../../../scripts/format_classpath.sh src.jar .`
 ^EXIT=10$
+^EXIT=10$
```
Comment (thk123) [inline jbmc/regression/jbmc/classpath-class-with-jar/test.desc]: Changing the tests at the same time as changing how they are opened makes this much harder to review

**Review by thk123 — COMMENTED** [review]
I got to the misnamed? commit `Make janalyzer use lazy goto model initialization` - please split up the changes to tests more logically to make reviewing easier (i.e. one commit where you just remove `.class` suffix, one commit where you add `EXIT=10` regex so can just be scanned through

File: jbmc/regression/jbmc/classpath-invalid/test-jar.desc [hunk 12]
```diff
@@ -1,10 +1,9 @@
...
 --classpath ./NotHere.jar
 ^EXIT=6$
 ^SIGNAL=0$
-^the program has no entry point$
-^failed to load class `Test'$
+Could not find or load main class Test
 --
 --
-symex-driven lazy loading should emit "the program has no entry point" but currently doesn't
+This is not quite 'java' behavior as 'java' always includes . in the classpath.
```
Comment (tautschnig) [inline jbmc/regression/jbmc/classpath-invalid/test-jar.desc:9]: Can't we just fix this?

File: jbmc/regression/jbmc/enum_switch/test.desc [hunk 13]
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-com/diffblue/regression/Foo.class
+com/diffblue/regression/Foo
```
Comment (tautschnig) [inline jbmc/regression/jbmc/enum_switch/test.desc]: Why does this one still use `/`  as a separator?

File: jbmc/regression/jbmc/enum_values_clone/test.desc [hunk 14]
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-com/diffblue/regression/EnumIter.class
+com/diffblue/regression/EnumIter
```
Comment (tautschnig) [inline jbmc/regression/jbmc/enum_values_clone/test.desc]: `.` instead of `/`?

File: jbmc/regression/jbmc/enum_values_clone_name/test.desc [hunk 15]
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-com/diffblue/regression/EnumIter.class
+com/diffblue/regression/EnumIter
```
Comment (tautschnig) [inline jbmc/regression/jbmc/enum_values_clone_name/test.desc]: `.` instead of `/`?

File: jbmc/regression/jbmc/internal1/test.desc [hunk 16]
```diff
@@ -1,5 +1,5 @@
 CORE
-com/diffblue/javatest/nestedobjects/subpackage/Order.class
+com/diffblue/javatest/nestedobjects/subpackage/Order
```
Comment (tautschnig) [inline jbmc/regression/jbmc/internal1/test.desc]: `.` instead of `/`?

File: jbmc/regression/jbmc/overlay-class/correct-test.desc [hunk 17]
```diff
@@ -1,13 +1,13 @@
...
+Test
 --classpath `../../../../scripts/format_classpath.sh . annotations correct-overlay` --verbosity 10
 ^Getting class `Test' from file \.[\\/]Test\.class$
 ^Getting class `Test' from file correct-overlay[\\/]Test\.class$
 ^Adding symbol from overlay class:  field 'x'$
 ^Adding symbol from overlay class:  method 'java::Test\.<init>:\(\)V'$
 ^Field definition for java::Test\.x already loaded from overlay class$
 ^Adding symbol from overlay class:  method 'java::Test\.main:\(\[Ljava/lang/String;\)V'$
-^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Comment (tautschnig) [inline jbmc/regression/jbmc/overlay-class/correct-test.desc]: Wrong auto-replacement?

File: jbmc/regression/jbmc/overlay-class/duplicate-test.desc [hunk 18]
```diff
@@ -8,7 +8,7 @@ Test.class
 ^Adding symbol from overlay class:  method 'java::Test\.<init>:\(\)V'$
 ^Field definition for java::Test\.x already loaded from overlay class$
 ^Adding symbol from overlay class:  method 'java::Test\.main:\(\[Ljava/lang/String;\)V'$
-^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Method java::Test\.<init>:\(\)V exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Comment (tautschnig) [inline jbmc/regression/jbmc/overlay-class/duplicate-test.desc]: Wrong auto-replacement?

File: jbmc/regression/jbmc/overlay-class/misordered-test.desc [hunk 19]
```diff
@@ -12,5 +12,5 @@ Test.class
 ^SIGNAL=0$
 --
 ^Skipping duplicate definition of class Test not marked with OverlayClassImplementation$
-^Adding symbol from overlay class
-exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Adding symbol from overlay
+exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Comment (tautschnig) [inline jbmc/regression/jbmc/overlay-class/misordered-test.desc:16]: Wrong auto-replacement?

File: jbmc/regression/jbmc/overlay-class/unmarked-test.desc [hunk 20]
```diff
@@ -12,5 +12,5 @@ Test.class
 ^SIGNAL=0$
 --
 ^Skipping class Test marked with OverlayClassImplementation but found before original definition$
-^Adding symbol from overlay class
-exists in an overlay class without being marked as an overlay and also exists in the underlying class
+^Adding symbol from overlay
+exists in an overlay class without being marked as an overlay and also exists in the underlying
```
Comment (tautschnig) [inline jbmc/regression/jbmc/overlay-class/unmarked-test.desc:16]: Wrong auto-replacement?

File: jbmc/regression/jbmc/provide_object_implementation/test.desc [hunk 21]
```diff
@@ -1,5 +1,5 @@
 CORE symex-driven-lazy-loading-expected-failure
-java/lang/Object.class
+java/lang/Object
```
Comment (tautschnig) [inline jbmc/regression/jbmc/provide_object_implementation/test.desc]: `.` instead of `/`?

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 22]
```diff
@@ -358,24 +362,77 @@ int janalyzer_parse_optionst::doit()
 
   register_languages();
 
-  if(cmdline.args.size() > 1)
+  if(!((cmdline.isset("jar") && cmdline.args.size() == 0) ||
+       (cmdline.isset("gb") && cmdline.args.size() == 0) ||
```
Comment (tautschnig) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: I'd prefer `.empty()` over `.size() == 0`

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 23]
```diff
@@ -358,24 +362,77 @@ int janalyzer_parse_optionst::doit()
 
   register_languages();
 
-  if(cmdline.args.size() > 1)
+  if(!((cmdline.isset("jar") && cmdline.args.size() == 0) ||
```
Comment (tautschnig) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: Could we push the negation inwards?

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 24]
```diff
@@ -358,24 +362,77 @@ int janalyzer_parse_optionst::doit()
...
+    CHECK_RETURN(language != nullptr);
+
+    language->set_language_options(options);
+    language->set_message_handler(ui_message_handler);
+
+    log.status() << "Parsing..." << messaget::eom;
 
-  if(process_goto_program(options))
+    const std::string filename = "";
+    std::ifstream infile(filename);
```
Comment (tautschnig) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: What is this about - opening a file without a name will surely fail?

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 25]
```diff
@@ -472,44 +472,49 @@ int jbmc_parse_optionst::doit()
   register_language(new_ansi_c_language);
   register_language(new_java_bytecode_language);
 
-  if(cmdline.isset("show-parse-tree"))
+  if(!((cmdline.isset("jar") && cmdline.args.size() == 0) ||
+       (cmdline.isset("gb") && cmdline.args.size() == 0) ||
```
Comment (tautschnig) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: As above: push negation inwards, use `.empty()`

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 26]
```diff
@@ -472,44 +472,49 @@ int jbmc_parse_optionst::doit()
...
+    CHECK_RETURN(language != nullptr);
 
     language->set_language_options(options);
     language->set_message_handler(ui_message_handler);
 
-    log.status() << "Parsing " << filename << messaget::eom;
+    log.status() << "Parsing..." << messaget::eom;
 
+    const std::string filename = "";
+    std::ifstream infile(filename);
```
Comment (tautschnig) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: As above - seems invalid.

File: jbmc/src/jbmc/jbmc_parse_options.h [hunk 27]
```diff
@@ -50,7 +50,10 @@ class optionst;
   "(debug-level):(no-propagation)(no-simplify-if)" \
   "(document-subgoals)(outfile):" \
   "(object-bits):" \
-  "(classpath):(cp):(main-class):" \
+  "(classpath):(cp):" \
+  "(jar):" \
+  "(main-class):" \
+  "(gb):" \
```
Comment (tautschnig) [inline jbmc/src/jbmc/jbmc_parse_options.h]: In line with @thk123's comment: this repeats, should be factored out.

File: regression/test.pl [hunk 28]
```diff
@@ -33,7 +33,12 @@ sub with_color {
...
+  if(length($input)) {
+    $cmdline = "$cmd $options '$input' >'$output' 2>&1";
+  } else {
+    $cmdline = "$cmd $options >'$output' 2>&1";
+  }
```
Comment (tautschnig) [inline regression/test.pl]: Does this actually address #3981?

File: jbmc/regression/jbmc/classpath-class-incorrect-package/test.desc (hunk 10 above)
Comment (peterschrammel) [reply jbmc/regression/jbmc/classpath-class-incorrect-package/test.desc:1]: This could go into a separate commit. 
Unfortunately all changes to the other tests need to go into the same commit together with changing the interface (including the one that is currently still split out).

File: jbmc/src/jbmc/jbmc_parse_options.cpp (hunk 4 above)
Comment (peterschrammel) [reply jbmc/src/jbmc/jbmc_parse_options.cpp]: Fully agreed. If it's ok to omit the program name from the help text then this seems be doable.

File: regression/test.pl (hunk 28 above)
Comment (peterschrammel) [reply regression/test.pl]: Yes, it does.

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 24 above)
Comment (peterschrammel) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: It doesn't fail, just gives a bad stream. It's ugly, though. I'll have to extend `languaget` to avoid this.

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 23 above)
Comment (peterschrammel) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: Hm, I actually prefer `not(or_i good_case_i)` to `and_i not(good_case_i)`

File: jbmc/regression/jbmc/classpath-class-with-jar/test.desc (hunk 11 above)
Comment (peterschrammel) [reply jbmc/regression/jbmc/classpath-class-with-jar/test.desc]: That's a superfluous change, actually.

File: regression/test.pl (hunk 28 above)
Comment (tautschnig) [reply regression/test.pl]: Cool! Please add the "Fixes: #3981" info to the commit message to have it auto-closed.

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 23 above)
Comment (tautschnig) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: I raised this, because the reader needs to 1) figure out matching parentheses and 2) read all the lines at once. Maybe you just want to introduce some named subexpressions?

File: src/goto-programs/lazy_goto_model.cpp (hunk 2 above)
Comment (peterschrammel) [reply src/goto-programs/lazy_goto_model.cpp]: > call some other languaget entry-point

I didn't want to open that can of worms... but since I have already started that work there's some hope that things change to the better... that will be a PR with a lot of noise through all language frontends...

**Comment by kroening** [comment]
`lazy_goto_modelt` is now really Java only, and should not live in goto-programs any more.

**Comment by peterschrammel** [comment]
> lazy_goto_modelt is now really Java only, and should not live in goto-programs any more.

@kroening, a solution that could work is to 
1. introduce in `languaget`
```
  virtual bool parse() = 0;
  virtual bool parse(const std::string &path) = 0;
```
to avoid having to supply dummy arguments.
2. deprecate
```
  virtual bool parse(std::istream &instream, const std::string &path)=0;
```
3. give (lazy_)goto_modelt initialize a default language to use with `parse()` (or use the language registry for that, e.g. 'the first language is the default' or similar). The implementation of `parse()` can then deal with the case where no input file is supplied and the language cannot handle that situation (which would then work transparently for C vs Java).

**Comment by kroening** [comment]
@peterschrammel This suffers from the following problem: `languaget` was intended to offer an interface that is common to multiple languages, i.e., really the interface that's the intersection.
The plan here is to make it the union, which isn't good practice.
I would suggest the following fix:
1) keep languaget as is, for the time being (that may be revisited, but seems independent of this PR)
2) jbmc is Java specific, as its name suggests; make jbmc talk to the Java frontend directly. If you like, introduce an interface for that to separate implementation from interface, but for now there would only be one provider.

The bottom line is that the idea to have a common language interface isn't helping if cbmc talks to C, and jbmc talks to Java. The code gets more complicated, not easier.

File: src/goto-programs/initialize_goto_model.cpp [hunk 29]
```diff
@@ -151,11 +152,21 @@ goto_modelt initialize_goto_model(
...
-    // existing __CPROVER_start function:
-    rebuild_goto_start_functiont rebuild_existing_start(
-      options, goto_model, msg.get_message_handler());
-    entry_point_generation_failed=rebuild_existing_start();
+    // Get the language annotation of the existing __CPROVER_start function.
+    std::unique_ptr<languaget> language = get_entry_point_language(
+      goto_model.symbol_table, options, message_handler);
+
+    // To create a new entry point we must first remove the old one
+    remove_existing_entry_point(goto_model.symbol_table);
```
Comment (romainbrenguier) [inline src/goto-programs/initialize_goto_model.cpp]: :bulb: would it have been possible to directly use the right entry point?

File: src/goto-programs/initialize_goto_model.cpp [hunk 30]
```diff
@@ -151,11 +152,21 @@ goto_modelt initialize_goto_model(
...
+
+    // Create the new entry-point
+    entry_point_generation_failed =
+      language->generate_support_functions(goto_model.symbol_table);
+
+    // Remove the function from the goto functions so it is copied back in
+    // from the symbol table during goto_convert
+    if(!entry_point_generation_failed)
```
Comment (romainbrenguier) [inline src/goto-programs/initialize_goto_model.cpp]: :question: is it safe to continue when it failed?

File: src/goto-programs/rebuild_goto_start_function.cpp [hunk 31]
```diff
@@ -30,8 +30,9 @@ std::unique_ptr<languaget> get_entry_point_language(
 {
   const irep_idt &mode = get_entry_point_mode(symbol_table);
 
-  // Get the relevant languaget to generate the new entry point with
+  // Get the relevant languaget to generate the new entry point with.
   std::unique_ptr<languaget> language = get_language_from_mode(mode);
+  // This might fail if the driver program hasn't registered that language.
   INVARIANT(language, "No language found for mode: " + id2string(mode));
```
Comment (romainbrenguier) [inline src/goto-programs/rebuild_goto_start_function.cpp]: :pick: maybe rephrase the invariant message so that is phrase as a "should" statement and includes as much  information as the comment (which can then be removed)

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 32]
```diff
@@ -701,6 +701,21 @@ void janalyzer_parse_optionst::help()
...
     "\n"
+    " -classpath dirs/jars\n"
+    " -cp dirs/jars\n"
+    " --classpath dirs/jars        set class search path of directories and\n"
+    "                              jar files\n"
+    "                              A "
+    #ifdef _WIN32
+    ";"
+    #else
+    ":"
```
Comment (romainbrenguier) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: :pick: this could be put in a `java_file_separator_char` function, as it can be used in several places of jbmc

File: jbmc/src/jbmc/jbmc_parse_options.h [hunk 33]
```diff
@@ -26,13 +26,13 @@ Author: Daniel Kroening, kroening@kroening.com
...
 #include <goto-programs/goto_trace.h>
 #include <goto-programs/show_properties.h>
-#include <java_bytecode/lazy_goto_model.h>
 
 #include <goto-symex/path_storage.h>
 
 #include <solvers/strings/string_refinement.h>
 
 #include <java_bytecode/java_bytecode_language.h>
+#include <java_bytecode/lazy_goto_model.h>
```
Comment (romainbrenguier) [inline jbmc/src/jbmc/jbmc_parse_options.h:35]: The "Move lazy_goto_modelt to JBMC" commit doesn't seem to match its description

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 34]
```diff
@@ -328,6 +315,45 @@ bool java_bytecode_languaget::parse(
 
     java_class_loader.set_extra_class_refs_function(get_string_base_classes);
   }
+}
+
+void java_bytecode_languaget::do_parse()
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp]: I don't think the name of the function describes well what it does

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 35]
```diff
@@ -375,6 +375,24 @@ int janalyzer_parse_optionst::doit()
     return CPROVER_EXIT_USAGE_ERROR;
   }
 
+  if(cmdline.args.size() == 1)
+  {
+    std::string main_class = cmdline.args[0];
+    std::replace(main_class.begin(), main_class.end(), '/', '.');
```
Comment (romainbrenguier) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp:380]: This could be clearer a bit clearer with an explanation or some good naming.

File: jbmc/src/java_bytecode/java_bytecode_language.cpp [hunk 36]
```diff
@@ -8,6 +8,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "java_bytecode_language.h"
 
+#include <fstream>
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_bytecode_language.cpp:11]: commit message doesn't seem to match its content

**Review by romainbrenguier — APPROVED** [review]

File: src/goto-programs/initialize_goto_model.cpp (hunk 30 above)
Comment (peterschrammel) [reply src/goto-programs/initialize_goto_model.cpp]: It will throw a `invalid_source_file_exceptiont` in line 191

File: src/goto-programs/initialize_goto_model.cpp (hunk 29 above)
Comment (peterschrammel) [reply src/goto-programs/initialize_goto_model.cpp]: The entry point is part of a goto-binary we loaded.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a1919cc2d72bd294801a8753ff7080dc0b6097a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/138039620
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 03af61633b767fff5e669338589ec6d99aa53998).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/138078934
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
